File: javaprolog/test_shrdlite_parser.pl
#!/usr/local/bin/swipl -g mainloop,halt -t 'halt(1)' -s

:- [dcg_parser].
:- [shrdlite_grammar].


mainloop :-
    nl, write('Write a sentence to parse; empty line or EOF to quit.'), nl, nl,
    repeat,
    read_line_to_codes(user_input, Codes),
    ( (Codes == end_of_file ; Codes == [])
    -> !,
       write('Goodbye!'), nl
    ;
       split_line(Codes, Words),
       write('Parsing: '), write(Words), nl,
       parse_all(command, Words, Trees),
       length(Trees, N),
       write(N), write(' solution(s)'), nl,
       forall(member(T, Trees),
              (write('--> '), write(T), nl)),
       nl,
       fail
    ).


%% split_line(+Codes : list(int), -Words : list(atom))
split_line(Codes, Words) :-
    atom_codes(Line, Codes),
    atomic_list_concat(Words0, ' ', Line),
    findall(W, (member(W, Words0), W \== ''), Words).

File: javaprolog/TestShrdliteParser.java

import java.util.Arrays;
import java.util.List;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;

import gnu.prolog.term.Term;
import gnu.prolog.term.CompoundTerm;
import gnu.prolog.vm.PrologException;

public class TestShrdliteParser {

	public static void main(String[] args) throws PrologException, IOException {
        System.out.println("Write a sentence to parse; empty line or EOF to quit.");

        DCGParser parser = new DCGParser("shrdlite_grammar.pl");
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String line;
        System.out.print("\n> ");
        while ((line = in.readLine()) != null && line.length() != 0) {
            String[] words = line.trim().split("\\s+");
            System.out.println("Parsing: " + Arrays.toString(words));
            List<ParseTree> results = parser.parseSentence("command", words);
            if (results.isEmpty()) {
                System.out.println("No solutions");
            } else {
                System.out.println(results.size() + " solution(s):");
                for (ParseTree t : results) {
                    System.out.println("NODE: " + t.getNode());
                    System.out.println("--> " + t);
                }
            }
            System.out.print("\n> ");
        }
        System.out.println("Goodbye!");
    }

}

